File: py/pytanga/viz/templates/views/toolbar-view.js
// Tanga Viewer — `ToolbarView`: a horizontal `StackView` with a thin border and
// a configurable inner margin (padding).  Sizes to its controls plus the
// border/padding "chrome" (see `_chrome`).

import { Size } from './size.js';
import { View } from './view.js';
import { StackView } from './stack-view.js';
import { stackCrossAxis, stackMinSize, stackPreferredSize } from './stack-size.js';

// Fallback chrome (border-box px per axis) used when the rendered DOM can't be
// measured (fake DOM in tests, or before the first layout).  Derived from the
// default `themes/views/toolbar-view.css` + inline margin:
//   1px border on each side + 6px margin on each side = 14px per axis.
const CHROME_FALLBACK = 14;

/**
 * A horizontal control toolbar: a bordered row of controls.
 * `direction` is fixed to `"horizontal"`; `margin` is the inner spacing
 * (padding) between the border and the controls; `border` toggles the outline;
 * `gap`/`align`/`justify` come straight from `StackView`.
 */
export class ToolbarView extends StackView {
    constructor({
        direction = 'horizontal',
        margin = null,
        border = true,
        gap = null,
        align = 'center',
        justify = 'start',
        children = [],
    } = {}) {
        super({ direction, gap, align, justify, children: [] });
        this.margin = Size.fromJSON(margin);
        this.border = border;

        this.el.classList.add('tanga-toolbar');
        if (border === false) this.el.classList.add('tanga-toolbar-borderless');
        this._applyMargin();

        // The cross axis (height) tracks the tallest rendered child rather than
        // the controls' min-height floors, so icon-only toolbars stay compact
        // and taller controls (dropdowns, sliders) don't overflow the chrome.
        this._contentHeightCache = null;
        this._contentObserver = null;
        this._onThemeChange = () => this._invalidateContent();
        if (typeof document !== 'undefined' && typeof document.addEventListener === 'function') {
            document.addEventListener('tanga:themechange', this._onThemeChange);
        }
        if (typeof ResizeObserver !== 'undefined') {
            this._contentObserver = new ResizeObserver(() => this._remeasureContent());
        }

        for (const child of children) this.addChild(child);
    }

    _applyMargin() {
        if (this.margin == null) return;
        const unit = this.margin.unit === '%' ? '%' : 'px';
        this.el.style.padding = `${this.margin.value}${unit}`;
    }

    // Measured padding + border along each axis (border-box px).  Falls back to
    // a constant when the DOM can't be measured (fake DOM in tests).
    _chrome() {
        const cs = typeof getComputedStyle === 'function' ? getComputedStyle(this.el) : null;
        if (!cs) return { x: CHROME_FALLBACK, y: CHROME_FALLBACK };
        const x = (parseFloat(cs.paddingLeft) || 0)
            + (parseFloat(cs.paddingRight) || 0)
            + (parseFloat(cs.borderLeftWidth) || 0)
            + (parseFloat(cs.borderRightWidth) || 0);
        const y = (parseFloat(cs.paddingTop) || 0)
            + (parseFloat(cs.paddingBottom) || 0)
            + (parseFloat(cs.borderTopWidth) || 0)
            + (parseFloat(cs.borderBottomWidth) || 0);
        return { x: Math.round(x), y: Math.round(y) };
    }

    // An explicit min/preferred describes the whole toolbar (chrome included),
    // so it is returned unchanged; otherwise the derived content size is bumped
    // by the border/padding on both axes.  The cross axis uses the measured
    // content height so the toolbar hugs its tallest control.
    minSizePx(axis, available) {
        const explicit = View.prototype.minSizePx.call(this, axis, available);
        const contentMin = axis === stackCrossAxis(this.direction)
            ? this._contentHeight(available)
            : stackMinSize(axis, this.direction, this.children, available);
        return Math.max(explicit, contentMin + this._chrome()[axis]);
    }

    preferredPx(axis, available) {
        const explicit = View.prototype.preferredPx.call(this, axis, available);
        if (explicit !== null && explicit !== undefined) return explicit;
        const contentPref = axis === stackCrossAxis(this.direction)
            ? this._contentHeight(available)
            : stackPreferredSize(axis, this.direction, this.children, available);
        return contentPref === null || contentPref === undefined
            ? null
            : contentPref + this._chrome()[axis];
    }

    // Fix the cross axis (height for a horizontal toolbar) to its natural
    // content height unless an explicit max is set.  This makes the toolbar a
    // fixed-height pane, so a vertical split collapses it to the toolbar's
    // height instead of stretching it.  The main axis stays unbounded so the
    // toolbar can still fill its pane horizontally.
    maxSizePx(axis, available) {
        const explicit = View.prototype.maxSizePx.call(this, axis, available);
        if (explicit !== null && explicit !== undefined) return explicit;
        if (axis === stackCrossAxis(this.direction)) {
            return this.minSizePx(axis, available);
        }
        return null;
    }

    // ── cross-axis content measurement ─────────────────────────

    // Measure the tallest child's rendered height (border-box px) along the
    // cross axis, or `null` when the DOM isn't measurable (fake DOM in tests,
    // or no laid-out children yet).
    _measureContent() {
        let max = 0;
        let measurable = false;
        for (const child of this.children) {
            const el = child.el;
            if (!el || typeof el.getBoundingClientRect !== 'function') continue;
            const rect = el.getBoundingClientRect();
            if (rect && rect.height > 0) {
                measurable = true;
                max = Math.max(max, rect.height);
            }
        }
        return measurable ? Math.round(max) : null;
    }

    // Cross-axis content height — the measured tallest child where possible,
    // otherwise the pure stack min (max of the children's min floors).  Cached;
    // invalidated on child add/remove/resize and theme change.
    _contentHeight(available) {
        if (this._contentHeightCache != null) return this._contentHeightCache;
        const measured = this._measureContent();
        this._contentHeightCache = measured !== null
            ? measured
            : stackMinSize(stackCrossAxis(this.direction), this.direction, this.children, available);
        return this._contentHeightCache;
    }

    // A child resized — drop the cached height and tell enclosing containers to
    // re-layout so they pick up the new value.
    _remeasureContent() {
        if (this._contentHeightCache == null) return;
        this._invalidateContent();
    }

    _invalidateContent() {
        this._contentHeightCache = null;
        this.emit('preferredchange', { fields: ['preferredWidth', 'preferredHeight'] });
        this.emit('constraintschange', { fields: ['minWidth', 'minHeight', 'maxWidth', 'maxHeight'] });
    }

    addChild(view) {
        this._contentHeightCache = null;
        const added = super.addChild(view);
        if (this._contentObserver && added.el) this._contentObserver.observe(added.el);
        this.emit('constraintschange', { fields: ['minWidth', 'minHeight', 'maxWidth', 'maxHeight'] });
        return added;
    }

    removeChild(view) {
        this._contentHeightCache = null;
        if (this._contentObserver && view.el) this._contentObserver.unobserve(view.el);
        super.removeChild(view);
        this.emit('constraintschange', { fields: ['minWidth', 'minHeight', 'maxWidth', 'maxHeight'] });
    }

    destroy() {
        if (this._contentObserver) this._contentObserver.disconnect();
        if (typeof document !== 'undefined' && typeof document.removeEventListener === 'function') {
            document.removeEventListener('tanga:themechange', this._onThemeChange);
        }
        super.destroy();
    }
}

File: py/pytanga/viz/views/toolbar_view.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2021 Christian Perwass

"""The :class:`ToolbarView` horizontal control toolbar."""

from __future__ import annotations

from typing import Any

from ._base import View
from .._controls import EControlVariant
from ._enums import EStackAlign, EStackDirection, EStackJustify
from ._helpers import _size_dict
from .._size import Size, SizeSpec
from .menu_view import MenuView
from .stack_view import StackView


def _apply_toolbar_variant(view: View) -> None:
    """Recursively force ``TOOLBAR`` onto eligible control views in a toolbar."""
    for child in getattr(view, "children", None) or ():
        if isinstance(child, MenuView):
            continue  # a nested menu keeps its own MENU styling
        ctrl = getattr(child, "control", None)
        if ctrl is not None and hasattr(ctrl, "variant"):
            ctrl.variant = EControlVariant.TOOLBAR
        _apply_toolbar_variant(child)


class ToolbarView(StackView):
    """A horizontal control toolbar (a bordered :class:`StackView` row).

    ``direction`` is fixed to ``"horizontal"``.  ``margin`` is the inner
    spacing (padding) between the border and the controls; ``border`` toggles
    the thin outline.  ``gap`` spaces the controls, ``align`` sets cross-axis
    (vertical) alignment, and ``justify`` positions the controls along the row
    (``START`` left, ``END`` right, ``CENTER`` block-centered, ``SPACE_EVENLY``
    equally spaced).
    """

    _node_type = "toolbar"

    def __init__(
        self,
        children: list[View] | None = None,
        *,
        margin: SizeSpec = Size.px(6),
        border: bool = True,
        gap: int | None = None,
        align: EStackAlign | str = EStackAlign.CENTER,
        justify: EStackJustify | str = EStackJustify.START,
        size: SizeSpec = None,
        preferred_width: SizeSpec = None,
        preferred_height: SizeSpec = None,
        min_width: SizeSpec = None,
        min_height: SizeSpec = None,
        max_width: SizeSpec = None,
        max_height: SizeSpec = None,
    ) -> None:
        if margin is not None and not isinstance(margin, Size):
            raise ValueError(f"margin must be a Size or None, got {margin!r}")
        if not isinstance(border, bool):
            raise ValueError(f"border must be a bool, got {border!r}")
        super().__init__(
            EStackDirection.HORIZONTAL,
            children,
            gap=gap,
            align=align,
            justify=justify,
            size=size,
            preferred_width=preferred_width,
            preferred_height=preferred_height,
            min_width=min_width,
            min_height=min_height,
            max_width=max_width,
            max_height=max_height,
        )
        self.margin = margin
        self.border = border
        _apply_toolbar_variant(self)

    def _serialize(self) -> dict[str, Any]:
        result = super()._serialize()
        result["margin"] = _size_dict(self.margin)
        result["border"] = self.border
        return result
